src/esp/bindings_js/modules/vr_demo.js

```diff
@@ -245,6 +279,87 @@ class VRDemo extends WebDemo {
     );
   }
 
+  handleInput(frame) {
+    for (let inputSource of frame.session.inputSources) {
+      if (inputSource.handedness != HANDEDNESS) {
+        continue;
+      }
+      if (inputSource.gamepad) {
+        const demo = this;
+
+        const fwdHeld = inputSource.gamepad.axes[FWD_AXIS] < -0.5;
+        if (fwdHeld && !this.prevFwdHeld) {
+          this.task.handleAction("moveForward");
+          window.setTimeout(function() {
+            demo.prevFwdHeld = false;
+          }, REPEAT_TIMEOUT);
+        }
+        this.prevFwdHeld = fwdHeld;
+
+        const leftHeld = inputSource.gamepad.axes[LEFT_RIGHT_AXIS] > 0.5;
+        if (leftHeld && !this.prevleftHeld) {
+          this.viewYawOffset += VIEW_YAW_STEP;
+          window.setTimeout(function() {
+            demo.prevLeftHeld = false;
+          }, REPEAT_TIMEOUT);
+        }
+        this.prevLeftHeld = leftHeld;
+
+        const rightHeld = inputSource.gamepad.axes[LEFT_RIGHT_AXIS] < -0.5;
+        if (rightHeld && !this.prevRightHeld) {
+          this.viewYawOffset -= VIEW_YAW_STEP;
+          window.setTimeout(function() {
+            demo.prevRightHeld = false;
+          }, REPEAT_TIMEOUT);
+        }
+        this.prevRightHeld = rightHeld;
+      }
+      /*
+      // 6DoF pose example
+      if (inputSource.gripSpace) {
+        const inputPose = frame.getPose(
+          inputSource.gripSpace,
+          this.xrReferenceSpace
+        );
+      }
+      */
+    }
+  }
+
+  updateHeadPose(pose, agent) {
+    const headRotation = Module.Quaternion.fromVec4f(
+      pointToArray(pose.transform.orientation)
+    );
+    const pointingVec = headRotation.transformVector(this.FWD);
+    const pointingAngle =
+      Math.atan2(pointingVec.z(), pointingVec.x()) - this.FWD_ANGLE;
+
+    const agentQuat = Module.Quaternion.rotation(
+      new Module.Rad(pointingAngle + this.viewYawOffset),
+      this.DOWN
+    );
+    const inverseAgentRot = Module.Quaternion.rotation(
+      new Module.Rad(-pointingAngle),
+      this.DOWN
+    );
+
+    agent.setRotation(agentQuat);
+
+    for (var iView = 0; iView < pose.views.length; ++iView) {
+      const view = pose.views[iView];
+
+      const sensor = agent.sensorSuite.get(VIEW_SENSORS[iView]);
+
+      const pos = pointToArray(view.transform.position).slice(0, -1); // don't need w for position
+      sensor.setLocalTransform(
```

Comment:
`setLocalTransform` takes a vec4f but you're passing a Quaternion. And the scary thing is that this builds and runs without error. I realize now that Emscripten/Embind/Javascript is not doing full type-checking. You can pass any object to `setLocalTransform` that happens to get storage as a 4-tuple of floats and you get no runtime error/warning.